File: Problems/No.1763_Longest_Nice_Substring/longest_nice_substring.py
class Solution(object):
    def longestNiceSubstring(self, s):
        """
        :type s: str
        :rtype: str
        """

        # parameter
        buddies = []

        # start computing
        # find buddies
        for c in s:
            if c.islower():
                buddies.append(c.upper())
            else:
                buddies.append(c.lower())
        
        # divide and conquer according to unfound buddies
        for i, c in enumerate(s):
            if c in buddies:
                continue
            left_substr = self.longestNiceSubstring(s[:i])
            right_substr = self.longestNiceSubstring(s[i+1:])
            return left_substr if (len(left_substr) >= len(right_substr)) else right_substr
        
        return s
        

File: Problems/No.1763_Longest_Nice_Substring/longest_nice_substring.cpp
class Solution {
public:
    string longestNiceSubstring(string s) {
        
        // parameter
        std::unordered_set<char> buddies;

        // start computing
        // find buddies
        for (char &c: s) {

            if (std::islower(c)) buddies.insert(std::toupper(c));
            else buddies.insert(std::tolower(c));

        }

        // divide and conquer according to unfound buddies
        for (int i=0; i<s.length(); i++) {

            if (buddies.count(s[i])) continue;

            std::string left_substr = longestNiceSubstring(s.substr(0, i));
            std::string right_substr = longestNiceSubstring(s.substr(i+1));

            return (left_substr.length() >= right_substr.length()) ? left_substr : right_substr;

        }

        return s;

    }
};
